File: Two Pointers/3Sum.py
"""
https://leetcode.com/problems/3sum/description/
"""

class Solution: #Time complexity O(n^2) #Space complexity O(1)
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        
        for i in range(len(nums)):
            if i > 0 and nums[i-1] == nums[i]:  #To avoid counting duplicates
                continue

            l = i + 1
            r = len(nums) - 1

            while l < r:
                threeSum = nums[i] + nums[l] + nums[r]

                if threeSum > 0:
                    r -= 1
                elif threeSum < 0:
                    l += 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    l += 1
                    while l < r and nums[l-1] == nums[l] : #To avoid counting duplicates
                        l += 1

        return res

File: Two Pointers/3Sum.js
/*
https://neetcode.io/problems/three-integer-sum/question
*/

class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     * Time compplexity O(n^2), Space complexity O(n) due to sorting
     */
    threeSum(nums) {
        nums.sort((a, b) => a - b)
        let res = []

        for(let i = 0; i < nums.length; i++){
            if(i > 0 && nums[i] === nums[i-1]){
                continue;
            }
            let j = i + 1
            let k = nums.length - 1

            while(j < k){
                let total = nums[i] + nums[j] + nums[k]

                if(total > 0){
                    k -= 1
                }else if(total < 0){
                    j += 1
                }else{
                    res.push([nums[i], nums[j], nums[k]])
                    j += 1
                    while(j < k && nums[j] === nums[j-1]){
                        j += 1
                    }
                }
            }
        }
        return res
    }
}
